File: src/main/java/me/chriss99/worldmanagement/Region.java
package me.chriss99.worldmanagement;

import org.joml.Vector2i;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.function.Function;

public class Region<T> {
    public final Vector2i coord;
    private final LinkedHashMap<Vector2i, T> tiles = new LinkedHashMap<>();

    public Region(Vector2i coord) {
        this.coord = coord;
    }

    public void addChunk(Vector2i coord, T chunk) {
        T oldTile = tiles.put(coord, chunk);
        if (oldTile != null)
            new IllegalStateException("Tile " + coord.x + ", " + coord.y+ " was overwritten!").printStackTrace();
    }

    public T getTile(Vector2i coord, Function<Vector2i, T> tileGenerator) {
        return tiles.computeIfAbsent(coord, tileGenerator);
    }

    public Set<Map.Entry<Vector2i, T>> getAllTiles() {
        return tiles.entrySet();
    }

    @Override
    public String toString() {
        return "Region{" +
                "coord=" + coord +
                ", tiles=" + tiles +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        Region<?> region = (Region<?>) o;
        return Objects.equals(coord, region.coord) && Objects.equals(tiles, region.tiles);
    }

    @Override
    public int hashCode() {
        return Objects.hash(coord, tiles);
    }
}

File: src/main/kotlin/worldmanagement/Region.kt
package me.chriss99.worldmanagement

import glm_.vec2.Vec2i
import java.util.*

class Region<T>(val coord: Vec2i) {
    private val tiles: LinkedHashMap<Vec2i, T> = LinkedHashMap<Vec2i, T>()
    val allTiles get() = tiles.entries.toList()

    fun addChunk(coord: Vec2i, chunk: T) {
        val oldTile = tiles.put(coord, chunk)
        if (oldTile != null)
            IllegalStateException("Tile " + coord.x + ", " + coord.y + " was overwritten!").printStackTrace()
    }

    operator fun get(coord: Vec2i, tileGenerator: (pos: Vec2i) -> T): T = tiles.computeIfAbsent(coord, tileGenerator)

    override fun toString() = "Region(coord=$coord, tiles=$tiles)"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Region<*>) return false

        if (coord != other.coord) return false
        if (tiles != other.tiles) return false

        return true
    }

    override fun hashCode() = Objects.hash(coord, tiles)
}